### PR TITLE
Add automated xUnit work item creation

### PR DESCRIPTION
**Pull request by jonfortescue**
This is the first pass attempt at auto-work item creation for xUnit projects using the Helix Job Sender. This should cover the majority of cross-repo use cases.

Note: this should not be merged until I finish the remaining work from [this](https://dev.azure.com/mseng/Tools/_git/CoreFX%20Engineering%20Infrastructure/pullrequest/403377?_a=overview) PR and merge it into OS onboarding as well.

cc / @markwilkie @Chrisboh @jaredpar @AdamYoblick @HaoK @natemcmaster @safern

### CODE REVIEW

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+            {
+                Log.LogError("Not all XUnit projects produced assemblies.");
+            }
+            else if (XUnitDllDirectories.Length < XUnitDllPaths.Length)
+            {
+                Log.LogError("More XUnit assemblies were found than projects which is alarming.");
+            }
+
+            DirectoriesToPathMap = new Dictionary<string, string>();
+            try
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Instead of throwing an exception if a match it found, consider a more performant solution like checking if(DirectoriesToPathMap.ContainsKey(...)) which runs in constant time.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+{
+    public class CreateXUnitWorkItems : Build.Utilities.Task
+    {
+        [Required]
+        public string[] XUnitDllDirectories { get; set; }
+
+        [Required]
+        public string[] XUnitDllPaths { get; set; }
+
+        private Dictionary<string, string> DirectoriesToPathMap;
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: private fields should be camelCase, not PascalCase

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 3]
```diff
@@ -0,0 +1,139 @@
+using System;
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Would be really really nice if there were XMLStyle (summary, etc...) comments on this code to explain what it's doing. For example, I don't understand how XUnitDllDirectories is related to XUnitDllPaths at first glance.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            try
+            {
+                for (int i = 0; i < XUnitDllDirectories.Length; i++)
+                {
+                    DirectoriesToPathMap.Add(XUnitDllDirectories[i], XUnitDllPaths[i]);
+                }
+            }
+            catch (ArgumentException e)
+            {
+                Log.LogError("Two identical publish paths were provided", e.StackTrace);
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Since your dictionary is case sensitive, two paths which are equal in windows will NOT cause the error you expect to happen here. Consider using a case-insensitive dictionary to force key collisions for strings that differ only by case.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+            {
+                Log.LogError("No XUnit Projects found");
+            }
+            else if (XUnitDllDirectories.Length > XUnitDllPaths.Length)
+            {
+                Log.LogError("Not all XUnit projects produced assemblies.");
+            }
+            else if (XUnitDllDirectories.Length < XUnitDllPaths.Length)
+            {
+                Log.LogError("More XUnit assemblies were found than projects which is alarming.");
```
Comment (MattGal) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: I know it's fun but can we please stop trying to be funny with logging messages?  "which is alarming" is really unnecessary.  Just state the error case.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 4 above)
Comment (MattGal) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: If I have mixed case paths on Windows I get past this limitation.  Should I worry?

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+            XUnitWorkItems = await Task.WhenAll(XUnitDllDirectories.Select(PrepareWorkItem));
+
+            return;
+        }
+
+        private async Task<ITaskItem> PrepareWorkItem(string publishPath)
+        {
+            await Task.Yield();
+
+            string assemblyName = Path.GetFileNameWithoutExtension(DirectoriesToPathMap[publishPath]);
```
Comment (MattGal) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Is there any case where the extension is not .dll and you're well served by pretending it is?

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+            string command = $"dotnet xunit.console.dll {assemblyName}.dll -xml testResults.xml";
+
+            Log.LogMessage($"Creating work item with properties Identity: {assemblyName}, PayloadDirectory: {publishPath}, Command: {command}");
+
+            ITaskItem workItem = new WiTaskItem(assemblyName, publishPath, command);
+            return workItem;
+        }
+
+        private class WiTaskItem : ITaskItem
```
Comment (MattGal) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Nit:  Consider a more descriptive name than Wi here.  It's just for the readability of a future person reading it.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        public string[] XUnitDllPaths { get; set; }
+
+        private Dictionary<string, string> DirectoriesToPathMap;
+
+        [Output]
+        public ITaskItem[] XUnitWorkItems { get; set; }
+
+        public override bool Execute()
+        {
+            if (XUnitDllDirectories.Length < 1)
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: nit: Even though XUnitDllDirectories is decorated with a Required attribute, it's still good practice to make sure reference types are not null before accessing them. (Because what if Required becomes optional at some point in the future?). The same is true for XUnitDllPaths.

Consider the following pattern at the top of the function:

`if (XUnitDllDirectories == null)
    Log.LogError(...)`

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+using System.Threading.Tasks;
+using Microsoft.Build.Framework;
+using Newtonsoft.Json;
+
+namespace Microsoft.DotNet.Helix.Sdk
+{
+    public class CreateXUnitWorkItems : Build.Utilities.Task
+    {
+        [Required]
+        public string[] XUnitDllDirectories { get; set; }
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: You should always use the most generic types when possible. Consider making these properties `IEnumerable<string>`, unless build tasks always use arrays, in which case ignore this comment.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+            return workItem;
+        }
+
+        private class WiTaskItem : ITaskItem
+        {
+            public WiTaskItem()
+            {
+                Metadata = new Dictionary<string, string>();
+            }
+            public WiTaskItem(string identity, string payloadDirectory, string command)
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: You should use constructor chaining here to avoid the duplicated line where you instantiate the Metadata dictionary. The syntax is as follows:

`public WiTaskItem(string identity, string payloadDirectory, string command) : this()`

Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Also you should always be validating your inputs before using them. Are you sure identity isn't null or empty? Etc...

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 11]
```diff
@@ -0,0 +1,139 @@
+using System;
+using System.Collections;
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Any unit tests? :)

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 1 above)
Comment (AdamYoblick) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: (I know the performance difference here is very minuscule, but throwing exceptions always comes with a performance hit. If you can resolve the error without doing that, which you can in this case, it's usually better to do so)

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 2 above)
Comment (jaredpar) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Also should be prefixed with an underscore.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 8 above)
Comment (jaredpar) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: What if this is null?

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+        private async Task ExecuteAsync()
+        {
+            XUnitWorkItems = await Task.WhenAll(XUnitDllDirectories.Select(PrepareWorkItem));
+
+            return;
+        }
+
+        private async Task<ITaskItem> PrepareWorkItem(string publishPath)
+        {
+            await Task.Yield();
```
Comment (jaredpar) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Why is this here?

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 6 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: No. This is purely for making the work item name pretty. If we'd rather leave the ".dll" in that we totally can.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 7 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Will do!

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 1 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: okay works for me!

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 4 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Or maybe not per my response comment there

Comment (AdamYoblick) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Probably yes, but it's safer to be more restrictive than not. The other option is to default the behavior to be case insensitive but allow the caller to pass in a boolean that allows for paths of the same name.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 12 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Async method requires an await handler to register as async.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+
+            string assemblyName = Path.GetFileNameWithoutExtension(_directoriesToPathMap[publishPath]);
+            string command = $"dotnet xunit.console.dll {assemblyName}.dll -xml testResults.xml";
```
Comment (safern) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: You're assuming that the build agent has dotnet installed in the path. In the current repos we use the one we bootstrap, should the task instead take a path to dotnet? Maybe as optional and if not provided default to "dotnet" ?? What if they're using a custom host passed as a correlationpayload?

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 14]
```diff
@@ -11,43 +11,58 @@ namespace Microsoft.DotNet.Helix.Sdk
...
+        /// An array of ITaskItems of type HelixWorkItem
+        /// </summary>
         [Output]
         public ITaskItem[] XUnitWorkItems { get; set; }
 
         public override bool Execute()
         {
+            if (XUnitDllDirectories is null || XUnitDllPaths is null)
+            {
+                Log.LogError($"Required metadata {(((XUnitDllDirectories is null) != (XUnitDllPaths is null)) ? (XUnitDllDirectories is null ? "XUnitDllDirectories is " : "XUnitDllPaths is ") : "XUnitDllDirectories and XUnitDllPaths are ")} null");
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Style: This is messy. The ternary operator usually makes code harder to read than not.

Consider putting each error in its own path:
`if (XUnitDllDirectories is null)
    ...
if (XUnitDllPaths is null)
    ...`

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Build.Framework;
+using Newtonsoft.Json;
+
+namespace Microsoft.DotNet.Helix.Sdk
+{
+    public class CreateXUnitWorkItems : Build.Utilities.Task
```
Comment (safern) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: It feels to me like all this could be done in a msbuild target rather than a custom task.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 13 above)
Comment (HaoK) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Related to this, ASP.NET would love the ability to specify both the sdk and runtime versions since ours aren't necessarily the same today, this is the issue I'm currently working on in our helix on boarding...

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 16]
```diff
@@ -11,43 +11,58 @@ namespace Microsoft.DotNet.Helix.Sdk
...
             }
             else if (XUnitDllDirectories.Length < XUnitDllPaths.Length)
             {
-                Log.LogError("More XUnit assemblies were found than projects which is alarming.");
+                Log.LogError("More XUnit assemblies were found than projects");
             }
 
-            DirectoriesToPathMap = new Dictionary<string, string>();
-            try
+            _directoriesToPathMap = new Dictionary<string, string>();
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Nit: execute should be executing the work item, not instantiating data structures. Consider moving this to the CreateXUnitWorkItems constructor.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 15 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Perhaps, but our current model is custom tasks for these kinds of things.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 13 above)
Comment (HaoK) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: We also have the added wrinkle that the versions we are building against aren't necessarily public versions either

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 17]
```diff
@@ -11,43 +11,58 @@ namespace Microsoft.DotNet.Helix.Sdk
...
             {
-                for (int i = 0; i < XUnitDllDirectories.Length; i++)
+                if (_directoriesToPathMap.ContainsKey(XUnitDllDirectories[i]))
                 {
-                    DirectoriesToPathMap.Add(XUnitDllDirectories[i], XUnitDllPaths[i]);
+                    Log.LogError("Two identical publish paths were provided");
+                }
+                else
+                {
+                    _directoriesToPathMap.Add(XUnitDllDirectories[i], XUnitDllPaths[i]);
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: This is not functionally equivalent to the old code. The old code was throwing an exception when a duplicate was found, which stopped it from processing the rest of the directories.

Just wanted to make sure you meant to change the logic.

Also, if two identical paths are provided, is this error meant to be terminal? Should you return false in that case?

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 18]
```diff
@@ -65,73 +80,72 @@ private async Task<ITaskItem> PrepareWorkItem(string publishPath)
...
             }
-            public WiTaskItem(string identity, string payloadDirectory, string command)
+            public WorkItemTaskItem(string identity, string payloadDirectory, string command) : this()
             {
-                Metadata = new Dictionary<string, string>();
-                ItemSpec = identity;
-                SetMetadata("Identity", identity);
-                SetMetadata("PayloadDirectory", payloadDirectory);
-                SetMetadata("Command", command);
+                ItemSpec = identity ?? "";
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: If identity, payload directory, or command are null or empty, you probably want to log an error or throw an ArgumentNullException. Just setting then to empty strings in the metadata doesn't seem like the right thing to do here.

Also, it's usually preferred to use string.Empty instead of ""

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 19]
```diff
@@ -65,73 +80,72 @@ private async Task<ITaskItem> PrepareWorkItem(string publishPath)
...
+                SetMetadata("Identity", identity ?? "");
+                SetMetadata("PayloadDirectory", payloadDirectory ?? "");
+                SetMetadata("Command", command ?? "");
             }
 
             public string ItemSpec { get; set; }
 
-            private Dictionary<string, string> Metadata;
+            private Dictionary<string, string> _metadata;
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: nit: fields and properties usually belong at the top of the class, followed by constructors, followed by methods.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 20]
```diff
@@ -65,73 +80,72 @@ private async Task<ITaskItem> PrepareWorkItem(string publishPath)
...
                 {
-                    destinationItem.SetMetadata(key, Metadata[key]);
+                    destinationItem.SetMetadata(key, _metadata[key]);
                 }
             }
 
             public string GetMetadata(string metadataName)
             {
-                return (Metadata.ContainsKey(metadataName) ? Metadata[metadataName] : "");
+                return (_metadata.ContainsKey(metadataName) ? _metadata[metadataName] : "");
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Trying to get metadata for a key that doesn't exist...I would think an error would be appropriate here instead of just returning empty string, but that could be a design decision you guys made.

Also string.Empty is usually preferred over ""

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 21]
```diff
@@ -65,73 +80,72 @@ private async Task<ITaskItem> PrepareWorkItem(string publishPath)
...
             }
 
             public void RemoveMetadata(string metadataName)
             {
-                if (Metadata.ContainsKey(metadataName))
+                if (_metadata.ContainsKey(metadataName))
                 {
-                    Metadata.Remove(metadataName);
+                    _metadata.Remove(metadataName);
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: You dont have to check if a dictionary contains a key to remove it. The call to remove returns true if something was removed, false if not.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 22]
```diff
@@ -65,73 +80,72 @@ private async Task<ITaskItem> PrepareWorkItem(string publishPath)
...
                 }
             }
 
             public void SetMetadata(string metadataName, string metadataValue)
             {
-                if (Metadata.ContainsKey(metadataName))
+                if (_metadata.ContainsKey(metadataName))
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: You don't have to check if the dictionary contains the key to add it. You can use the indexer to add/set all at once. Just `_metadata[metadataName] = metadataValue;` works in both cases

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 23]
```diff
@@ -11,43 +11,58 @@ namespace Microsoft.DotNet.Helix.Sdk
 {
     public class CreateXUnitWorkItems : Build.Utilities.Task
     {
+        /// <summary>
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Missing xml documentation on the class.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 24]
```diff
@@ -11,43 +11,58 @@ namespace Microsoft.DotNet.Helix.Sdk
...
-        private Dictionary<string, string> DirectoriesToPathMap;
+        private Dictionary<string, string> _directoriesToPathMap;
 
+        /// <summary>
+        /// An array of ITaskItems of type HelixWorkItem
+        /// </summary>
         [Output]
         public ITaskItem[] XUnitWorkItems { get; set; }
 
         public override bool Execute()
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Missing xml documentation on the methods. Please make sure to include parameters (if there are any) and returns as well. :)

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 4 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Done

Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Resolved forreal now.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 11 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: My plan for now is to move forward with this PR without adding unit tests, since we don't have a full plan for how we want to test this package in general. We do have an epic open for adding tests to the Arcade SDK and I've filed an [issue](https://github.com/dotnet/arcade/issues/1313) under that epic to track this work.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 13 above)
Comment (HaoK) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: @jonfortescue sure do you want me to start an email or file an issue with some of the additional wrinkles we have?

Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: @HaoK if you could send an email with some of the particulars for your case I think that'd be very helpful!

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 17 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Log.LogError will ultimately return false on this regardless.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 25]
```diff
@@ -0,0 +1,164 @@
...
+            string dotNetPath = string.IsNullOrEmpty(PathToDotnet) ? "dotnet" : PathToDotnet;
+            string command = $"{dotNetPath} xunit.console.dll {assemblyName}.dll -xml testResults.xml";
+
+            Log.LogMessage($"Creating work item with properties Identity: {assemblyName}, PayloadDirectory: {publishPath}, Command: {command}");
+
+            ITaskItem workItem = new WorkItemTaskItem(assemblyName, publishPath, command);
+            return workItem;
+        }
+
+        private class WorkItemTaskItem : ITaskItem
```
Comment (alexperovich) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Is there a reason you created your own implementation of ITaskItem? IIRC msbuild has the implementation type `TaskItem` that you can use.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs (hunk 12 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: Added a comment for clarity.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+                Log.LogError("Required metadata XUnitDllPaths is null");
+                return;
+            }
+
+            if (XUnitDllDirectories.Length < 1)
+            {
+                Log.LogError("No XUnit Projects found");
+                return;
+            }
+            else if (XUnitDllDirectories.Length > XUnitDllPaths.Length)
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: nit: The previous if has a return in it, meaning an else is not needed here. This can just be an if.

File: src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+            {
+                Log.LogError("Not all XUnit projects produced assemblies");
+                return;
+            }
+            else if (XUnitDllDirectories.Length < XUnitDllPaths.Length)
```
Comment (AdamYoblick) [inline src/Microsoft.DotNet.Helix/Sdk/CreateXUnitWorkItems.cs]: nit: same as above

File: eng/common/templates/steps/helix-publish.yml [hunk 28]
```diff
@@ -9,6 +9,9 @@ parameters:
   WorkItemDirectory: ''
   WorkItemCommand: ''
   CorrelationPayloadDirectory: ''
+  XUnitProjects: ''
```
Comment (chcosta) [inline eng/common/templates/steps/helix-publish.yml]: I wonder if it's cleaner to turn these parameters into objects rather than strings.  ie…

```
parameters:
  DotNetCli: {}
  Helix: {}
  WorkItem: {}
  XUnit: {}
  CorrelationPayloadDirectory: ''
  EnableXUnitReporter: false
  IncludeDotNetCli: false
  WaitForWorkItemCompletion: true
  condition: succeeded()
  continueOnError: false
```

Then you define the schema for those objects in some [doc like Azure DevOps does](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=vsts&tabs=schema#pipeline) and link to it here.  That would reduce the gobs of parameters that make this a daunting template.

**Comment by jonfortescue** [comment]
nupkg unzipping tested and working in int: https://mc.int-dot.net/#/user/jonfortescue/pr~2Fjonfortescue/test/13112018.5/workItem/test/wilogs

File: eng/common/templates/steps/helix-publish.yml (hunk 28 above)
Comment (jonfortescue) [reply eng/common/templates/steps/helix-publish.yml]: That's a good idea. I'll look into cleaning it up that way in a future PR.

**Comment by jonfortescue** [comment]
And we have a working test of this in int:
https://mc.int-dot.net/#/user/jonfortescue/pr~2Fjonfortescue/test/13112018.22/workItem/test/wilogs
